File: src/commands.js
// mautrix-telegram - A Matrix-Telegram puppeting bridge
// Copyright (C) 2017 Tulir Asokan
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
const makePasswordHash = require("telegram-mtproto").plugins.makePasswordHash
const Portal = require("./portal")

const commands = {}

/**
 * Module containing all management commands.
 *
 * @module commands
 */

/**
 * Run management command.
 *
 * @param {string}          sender  The MXID of the user who sent the command.
 * @param {string}          command The command itself.
 * @param {Array<string>}   args    A list of arguments.
 * @param {function}        reply   A function that is called to reply to the command.
 * @param {object}          extra   Extra information that the handlers may find useful.
 * @param {MautrixTelegram} extra.app          The app main class instance.
 * @param {MatrixEvent}     extra.evt          The event that caused this call.
 * @param {string}          extra.roomID       The ID of the Matrix room the command was sent to.
 * @param {boolean}         extra.isManagement Whether or not the Matrix room is a management room.
 * @param {boolean}         extra.isPortal     Whether or not the Matrix room is a portal to a Telegram chat.
 */
function run(sender, command, args, reply, extra) {
	const commandFunc = this.commands[command]
	if (!commandFunc) {
		if (sender.commandStatus) {
			if (command === "cancel") {
				reply(`${sender.commandStatus.action} cancelled.`)
				sender.commandStatus = undefined
				return undefined
			}
			args.unshift(command)
			return sender.commandStatus.next(sender, args, reply, extra)
		}
		reply("Unknown command. Try `$cmdprefix help` for help.")
		return undefined
	}
	try {
		return commandFunc(sender, args, reply, extra)
	} catch (err) {
		reply(`Error running command: ${err}.`)
		if (err instanceof Error) {
			reply(["```", err.stack, "```"].join(""))
			console.error(err.stack)
		}
	}
	return undefined
}

commands.cancel = () => "Nothing to cancel."

commands.help = (sender, args, reply, { isManagement, isPortal }) => {
	let replyMsg = ""
	if (isManagement) {
		replyMsg += "This is a management room: prefixing commands with `$cmdprefix` is not required.\n"
	} else if (isPortal) {
		replyMsg += "**This is a portal room**: you must always prefix commands with `$cmdprefix`.\n" +
			"Management commands will not be sent to Telegram.\n"
	} else {
		replyMsg += "**This is not a management room**: you must prefix commands with `$cmdprefix`.\n"
	}
	replyMsg += `
_**Generic bridge commands**: commands for using the bridge that aren't related to Telegram._<br/>
**help** - Show this help message.<br/>
**cancel** - Cancel an ongoing action (such as login).<br/>
**setManagement** - Mark the room as a management room.<br/>
**unsetManagement** - Undo management room marking.

_**Telegram actions**: commands for using the bridge to interact with Telegram._<br/>
**login** &lt;_phone_&gt; - Request an authentication code.<br/>
**logout** - Log out from Telegram.<br/>
**search** [_-r|--remote_] &lt;_query_&gt; - Search your contacts or the Telegram servers for users.<br/>
**create** &lt;_group/channel_&gt; [_room ID_] - Create a Telegram chat of the given type for a Matrix room.
                                           If the room ID is not specified, a chat for the current room is created.<br/>
**upgrade** - Upgrade a normal Telegram group to a supergroup.

_**Temporary commands**: commands that will be replaced with more Matrix-y actions later._<br/>
**pm** &lt;_id_&gt; - Open a private chat with the given Telegram user ID.

_**Debug commands**: commands to help in debugging the bridge. Disabled by default._<br/>
**api** &lt;_method_&gt; &lt;_args_&gt; - Call a Telegram API method. Args is always a single JSON object.
`
	reply(replyMsg, { allowHTML: true })
}

commands.setManagement = (sender, _, reply, { app, roomID, isPortal }) => {
	if (isPortal) {
		reply("You may not mark portal rooms as management rooms.")
		return
	}
	app.managementRooms.push(roomID)
	reply("Room marked as management room. You can now run commands without the `$cmdprefix` prefix.")
}

commands.unsetManagement = (sender, _, reply, { app, roomID }) => {
	app.managementRooms.splice(app.managementRooms.indexOf(roomID), 1)
	reply("Room unmarked as management room. You must now include the `$cmdprefix` prefix when running commands.")
}


/////////////////////////////
// Authentication handlers //
/////////////////////////////

/**
 * Two-factor authentication handler.
 */
commands.enterPassword = async (sender, args, reply, { isManagement }) => {
	if (!isManagement) {
		reply("Logging in is considered a confidential action, and thus is only allowed in management rooms.")
		return
	} else if (args.length === 0) {
		reply("**Usage:** `$cmdprefix <password> [salt]`")
		return
	}

	let salt

	if (!sender.commandStatus || !sender.commandStatus.salt) {
		if (args.length > 1) {
			salt = args[1]
		} else {
			reply("No password salt found. Did you enter your phone code already?")
			return
		}
	} else {
		salt = sender.commandStatus.salt
	}

	const hash = makePasswordHash(salt, args[0])
	try {
		await sender.checkPassword(hash)
		reply(`Logged in successfully as ${sender.telegramPuppet.getDisplayName()}.`)
		sender.commandStatus = undefined
	} catch (err) {
		reply(`**Login failed:** ${err}`)
		if (err instanceof Error) {
			reply(["```", err.stack, "```"].join(""))
			console.error(err.stack)
		}
	}
}

/*
 * Login code send handler.
 */
commands.enterCode = async (sender, args, reply, { isManagement }) => {
	if (!isManagement) {
		reply("Logging in is considered a confidential action, and thus is only allowed in management rooms.")
		return
	} else 	if (args.length === 0) {
		reply("**Usage:** `$cmdprefix <authentication code>`")
		return
	}

	try {
		const data = await sender.signInToTelegram(args[0])
		if (data.status === "ok") {
			reply(`Logged in successfully as ${sender.telegramPuppet.getDisplayName()}.`)
			sender.commandStatus = undefined
		} else if (data.status === "need-password") {
			reply(`You have two-factor authentication enabled. Password hint: ${data.hint}
Enter your password using \`$cmdprefix <password>\``)
			sender.commandStatus = {
				action: "Two-factor authentication",
				next: commands.enterPassword,
				salt: data.salt,
			}
		} else {
			reply(`Unexpected sign in response, status=${data.status}`)
		}
	} catch (err) {
		reply(`**Login failed:** ${err}`)
		if (err instanceof Error) {
			reply(["```", err.stack, "```"].join(""))
			console.error(err.stack)
		}
	}
}

/*
 * Login code request handler.
 */
commands.login = async (sender, args, reply, { isManagement }) => {
	if (!isManagement) {
		reply("Logging in is considered a confidential action, and thus is only allowed in management rooms.")
		return
	} else if (args.length === 0) {
		reply("**Usage:** `$cmdprefix login <phone number>`")
		return
	}

	try {
		/*const data = */
		await sender.sendTelegramCode(args[0])
		reply(`Login code sent to ${args[0]}.\nEnter the code using \`$cmdprefix <code>\``)
		sender.commandStatus = {
			action: "Phone code authentication",
			next: commands.enterCode,
		}
	} catch (err) {
		reply(`**Failed to send code:** ${err}`)
		if (err instanceof Error) {
			reply(["```", err.stack, "```"].join(""))
			console.error(err.stack)
		}
	}
}

commands.register = async (sender, args, reply) => {
	reply("Registration has not yet been implemented. Please use the official apps for now.")
}

commands.logout = async (sender, args, reply) => {
	try {
		sender.logOutFromTelegram()
		reply("Logged out successfully.")
	} catch (err) {
		reply(`**Failed to log out:** ${err}`)
		if (err instanceof Error) {
			reply(["```", err.stack, "```"].join(""))
			console.error(err.stack)
		}
	}
}

//////////////////////////////
// General command handlers //
//////////////////////////////

commands.create = async (sender, args, reply, { app, roomID }) => {
	if (args.length < 1 || (args[0] !== "group" && args[0] !== "channel")) {
		reply("**Usage:** `$cmdprefix create <group/channel>`")
		return
	} else if (!sender._telegramPuppet) {
		reply("This command requires you to be logged in.")
		return
	} else if (args[0] === "channel") {
		reply("Creating channels is not yet supported.")
		return
	}

	if (args.length > 1) {
		roomID = args[1]
	}

	// TODO make sure that the AS bot is in the room.

	const title = await app.getRoomTitle(roomID)
	if (!title) {
		reply("Please set a room name before creating a Telegram chat.")
		return
	}

	let portal = await app.getPortalByRoomID(roomID)
	if (portal) {
		reply("This is already a portal room.")
		return
	}

	portal = new Portal(app, roomID)
	try {
		await portal.createTelegramChat(sender.telegramPuppet, title)
		reply(`Telegram chat created. ID: ${portal.id}`)
		if (app.managementRooms.includes(roomID)) {
			app.managementRooms.splice(app.managementRooms.indexOf(roomID), 1)
		}
	} catch (err) {
		reply(`Failed to create Telegram chat: ${err}`)
	}
}

commands.upgrade = async (sender, args, reply, { app, roomID }) => {
	if (!sender._telegramPuppet) {
		reply("This command requires you to be logged in.")
		return
	}

	const portal = await app.getPortalByRoomID(roomID)
	if (!portal) {
		reply("This is not a portal room.")
		return
	}

	await portal.upgradeTelegramChat(sender.telegramPuppet)
}

commands.search = async (sender, args, reply, { app }) => {
	if (args.length < 1) {
		reply("**Usage:** `$cmdprefix search [-r|--remote] <query>`")
		return
	} else if (!sender._telegramPuppet) {
		reply("This command requires you to be logged in.")
		return
	}
	const msg = []
	if (args[0] !== "-r" && args[0] !== "--remote") {
		const contactResults = await sender.searchContacts(args.join(" "))
		if (contactResults.length > 0) {
			msg.push("**Following results found from local contacts:**")
			msg.push("")
			for (const { match, contact } of contactResults) {
				msg.push(`- <a href="${
					app.getMatrixToLinkForTelegramUser(contact.id)}">${contact.getDisplayName()}</a>: ${contact.id} (${match}% match)`)
			}
			msg.push("")
			msg.push("To force searching from Telegram servers, add `-r` before the search query.")
			reply(msg.join("\n"), { allowHTML: true })
			return
		}
	} else {
		args.shift()
		msg.push("-r flag found: forcing remote search")
		msg.push("")
	}
	const query = args.join(" ")
	if (query.length < 5) {
		reply("Failed to search server: Query is too short.")
		return
	}
	const telegramResults = await sender.searchTelegram(query)
	if (telegramResults.length > 0) {
		msg.push("**Following results received from Telegram server:**")
		for (const user of telegramResults) {
			msg.push(`- <a href="${
				app.getMatrixToLinkForTelegramUser(user.id)}">${user.getDisplayName()}</a>: ${user.id}`)
		}
	} else {
		msg.push("**No users found.**")
	}
	reply(msg.join("\n"), { allowHTML: true })
}

commands.pm = async (sender, args, reply, { app }) => {
	if (args.length < 1) {
		reply("**Usage:** `$cmdprefix pm <id>`")
		return
	} else if (!sender._telegramPuppet) {
		reply("This command requires you to be logged in.")
		return
	}
	const user = await app.getTelegramUser(+args[0], { createIfNotFound: false })
	if (!user) {
		reply("User info not saved. Try searching for the user first?")
		return
	}
	const peer = user.toPeer(sender.telegramPuppet)

	const userInfo = await peer.getInfo(sender.telegramPuppet)
	await user.updateInfo(sender.telegramPuppet, userInfo)

	const portal = await app.getPortalByPeer(peer)
	await portal.createMatrixRoom(sender.telegramPuppet, {
		invite: [sender.userID],
	})
}

////////////////////////////
// Debug command handlers //
////////////////////////////

commands.api = async (sender, args, reply, { app }) => {
	if (!app.config.bridge.commands.allow_direct_api_calls) {
		reply("Direct API calls are forbidden on this mautrix-telegram instance.")
		return
	}
	const apiMethod = args.shift()
	let apiArgs
	try {
		apiArgs = JSON.parse(args.join(" "))
	} catch (err) {
		reply("Invalid API method parameters. Usage: $cmdprefix api <method> <json data>")
		return
	}
	try {
		reply(`Calling ${apiMethod} with the following arguments:\n${JSON.stringify(apiArgs, "", "  ")}`)
		const response = await sender.telegramPuppet.client(apiMethod, apiArgs)
		reply(`API call successful. Response:

<pre><code class="language-json">
  ${JSON.stringify(response, "", "  ")}
</code></pre>`, { allowHTML: true })
	} catch (err) {
		reply(`API call errored. Response:\n${JSON.stringify(err, "", "  ")}`)
	}
}

function timeout(promise, ms = 2500) {
	return new Promise((resolve, reject) => {
		promise.then(resolve, reject)
		setTimeout(() => reject(new Error("API call response not received")), ms)
	})
}

commands.ping = async (sender, args, reply) => {
	try {
		await timeout(sender.telegramPuppet.client("contacts.getContacts", {}))
		reply("Connection seems OK.")
	} catch (err) {
		reply(`Not connected: ${err}`)
	}
}

module.exports = {
	commands,
	run,
}

File: mautrix_telegram/commands.py
# -*- coding: future_fstrings -*-
# mautrix-telegram - A Matrix-Telegram puppeting bridge
# Copyright (C) 2018 Tulir Asokan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
import markdown
import logging
import asyncio

from mautrix_appservice import MatrixRequestError

from telethon.errors import *
from telethon.tl.types import *
from telethon.tl.functions.messages import ImportChatInviteRequest, CheckChatInviteRequest
from telethon.tl.functions.channels import JoinChannelRequest

from . import puppet as pu, portal as po

command_handlers = {}


def command_handler(func):
    command_handlers[func.__name__] = func
    return func


def format_duration(seconds):
    def pluralize(count, singular): return singular if count == 1 else singular + "s"

    def include(count, word): return f"{count} {pluralize(count, word)}" if count > 0 else ""

    minutes, seconds = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    days, hours = divmod(hours, 24)
    parts = [a for a in [
        include(days, "day"),
        include(hours, "hour"),
        include(minutes, "minute"),
        include(seconds, "second")] if a]
    if len(parts) > 2:
        return "{} and {}".format(", ".join(parts[:-1]), parts[-1])
    return " and ".join(parts)


class CommandEvent:
    def __init__(self, az, command_prefix, room, sender, args, is_management, is_portal):
        self.az = az
        self.command_prefix = command_prefix
        self.room_id = room
        self.sender = sender
        self.args = args
        self.is_management = is_management
        self.is_portal = is_portal

    def reply(self, message, allow_html=False, render_markdown=True):
        if not self.room_id:
            raise AttributeError("the reply function can only be used from within"
                                 "the `CommandHandler.run` context manager")

        message = message.replace("$cmdprefix+sp ",
                                  "" if self.is_management else f"{self.command_prefix} ")
        message = message.replace("$cmdprefix", self.command_prefix)
        html = None
        if render_markdown:
            html = markdown.markdown(message, safe_mode="escape" if allow_html else False)
        elif allow_html:
            html = message
        return self.az.intent.send_notice(self.room_id, message, html=html)


class CommandHandler:
    log = logging.getLogger("mau.commands")

    def __init__(self, context):
        self.az, self.db, self.config, self.loop = context
        self.command_prefix = self.config["bridge.command_prefix"]

    # region Utility functions for handling commands

    async def handle(self, room, sender, command, args, is_management, is_portal):
        evt = CommandEvent(self.az, self.command_prefix, room, sender, args, is_management,
                           is_portal)
        command = command.lower()
        try:
            command = command_handlers[command]
        except KeyError:
            if sender.command_status and "next" in sender.command_status:
                args.insert(0, command)
                command = sender.command_status["next"]
            else:
                command = command_handlers["unknown_command"]
        try:
            await command(self, evt)
        except FloodWaitError as e:
            return evt.reply(f"Flood error: Please wait {format_duration(e.seconds)}")
        except Exception:
            self.log.exception(f"Fatal error handling command "
                               + f"'$cmdprefix {command} {''.join(args)}' from {sender.mxid}")
            return evt.reply("Fatal error while handling command. Check logs for more details.")

    # endregion
    # region Command handlers

    @command_handler
    async def ping(self, evt):
        if not evt.sender.logged_in:
            return await evt.reply("You're not logged in.")
        me = await evt.sender.client.get_me()
        if me:
            return await evt.reply(f"You're logged in as @{me.username}")
        else:
            return await evt.reply("You're not logged in.")

    # region Authentication commands
    @command_handler
    def register(self, evt):
        return evt.reply("Not yet implemented.")

    @command_handler
    async def login(self, evt):
        if not evt.is_management:
            return await evt.reply(
                "`login` is a restricted command: you may only run it in management rooms.")
        elif evt.sender.logged_in:
            return await evt.reply("You are already logged in.")
        elif len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp login <phone number>`")
        phone_number = evt.args[0]
        await evt.sender.client.sign_in(phone_number)
        evt.sender.command_status = {
            "next": command_handlers["enter_code"],
            "action": "Login",
        }
        return await evt.reply(f"Login code sent to {phone_number}. Please send the code here.")

    @command_handler
    async def enter_code(self, evt):
        if not evt.sender.command_status:
            return await evt.reply(
                "Request a login code first with `$cmdprefix+sp login <phone>`")
        elif len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp enter_code <code>`")

        try:
            user = await evt.sender.client.sign_in(code=evt.args[0])
            asyncio.ensure_future(evt.sender.post_login(user), loop=self.loop)
            evt.sender.command_status = None
            return await evt.reply(f"Successfully logged in as @{user.username}")
        except PhoneNumberUnoccupiedError:
            return await evt.reply("That phone number has not been registered."
                                   "Please register with `$cmdprefix+sp register <phone>`.")
        except PhoneCodeExpiredError:
            return await evt.reply(
                "Phone code expired. Try again with `$cmdprefix+sp login <phone>`.")
        except PhoneCodeInvalidError:
            return await evt.reply("Invalid phone code.")
        except PhoneNumberAppSignupForbiddenError:
            return await evt.reply(
                "Your phone number does not allow 3rd party apps to sign in.")
        except PhoneNumberFloodError:
            return await evt.reply(
                "Your phone number has been temporarily blocked for flooding. "
                "The block is usually applied for around a day.")
        except PhoneNumberBannedError:
            return await evt.reply("Your phone number has been banned from Telegram.")
        except SessionPasswordNeededError:
            evt.sender.command_status = {
                "next": command_handlers["enter_password"],
                "action": "Login (password entry)",
            }
            return await evt.reply("Your account has two-factor authentication."
                                   "Please send your password here.")
        except Exception:
            self.log.exception("Error sending phone code")
            return await evt.reply("Unhandled exception while sending code."
                                   "Check console for more details.")

    @command_handler
    async def enter_password(self, evt):
        if not evt.sender.command_status:
            return await evt.reply(
                "Request a login code first with `$cmdprefix+sp login <phone>`")
        elif len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp enter_password <password>`")

        try:
            user = await evt.sender.client.sign_in(password=evt.args[0])
            asyncio.ensure_future(evt.sender.post_login(user), loop=self.loop)
            evt.sender.command_status = None
            return await evt.reply(f"Successfully logged in as @{user.username}")
        except PasswordHashInvalidError:
            return await evt.reply("Incorrect password.")
        except Exception:
            self.log.exception("Error sending password")
            return await evt.reply("Unhandled exception while sending password. "
                                   "Check console for more details.")

    @command_handler
    async def logout(self, evt):
        if not evt.sender.logged_in:
            return await evt.reply("You're not logged in.")
        if await evt.sender.log_out():
            return await evt.reply("Logged out successfully.")
        return await evt.reply("Failed to log out.")

    # endregion
    # region Telegram interaction commands

    @command_handler
    async def search(self, evt):
        if len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp search [-r|--remote] <query>`")
        elif not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        force_remote = False
        if evt.args[0] in {"-r", "--remote"}:
            force_remote = True
            evt.args.pop(0)

        query = " ".join(evt.args)
        if force_remote and len(query) < 5:
            return await evt.reply("Minimum length of query for remote search is 5 characters.")

        results, remote = await evt.sender.search(query, force_remote)

        reply = []
        if remote:
            reply += ["**Results from Telegram server:**", ""]
        else:
            reply += ["**Results in contacts:**", ""]
        reply += [(f"* [{puppet.displayname}](https://matrix.to/#/{puppet.mxid}): "
                   + f"{puppet.id} ({similarity}% match)")
                  for puppet, similarity in results]

        # TODO somehow show remote channel results when joining by alias is possible?

        return await evt.reply("\n".join(reply))

    @command_handler
    async def pm(self, evt):
        if len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp pm <user identifier>`")
        elif not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        user = await evt.sender.client.get_entity(evt.args[0])
        if not user:
            return await evt.reply("User not found.")
        elif not isinstance(user, User):
            return await evt.reply("That doesn't seem to be a user.")
        portal = po.Portal.get_by_entity(user, evt.sender.tgid)
        await portal.create_matrix_room(evt.sender, user, [evt.sender.mxid])
        return await evt.reply(
            f"Created private chat room with {pu.Puppet.get_displayname(user, False)}")

    @command_handler
    async def invitelink(self, evt):
        if not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        portal = po.Portal.get_by_mxid(evt.room_id)
        if not portal:
            return await evt.reply("This is not a portal room.")

        if portal.peer_type == "user":
            return await evt.reply("You can't invite users to private chats.")

        try:
            link = await portal.get_invite_link(evt.sender)
            return await evt.reply(f"Invite link to {portal.title}: {link}")
        except ValueError as e:
            return await evt.reply(e.args[0])
        except ChatAdminRequiredError:
            return await evt.reply("You don't have the permission to create an invite link.")

    @command_handler
    async def deleteportal(self, evt):
        if not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")
        elif not evt.sender.is_admin:
            return await evt.reply("This is command requires administrator privileges.")

        portal = po.Portal.get_by_mxid(evt.room_id)
        if not portal:
            return await evt.reply("This is not a portal room.")

        for user in await portal.main_intent.get_room_members(portal.mxid):
            if user != portal.main_intent.mxid:
                try:
                    await portal.main_intent.kick(portal.mxid, user, "Portal deleted.")
                except MatrixRequestError:
                    pass
        await portal.main_intent.leave_room(portal.mxid)
        portal.delete()

    @staticmethod
    def _strip_prefix(value, prefixes):
        for prefix in prefixes:
            if value.startswith(prefix):
                return value[len(prefix):]
        return value

    @command_handler
    async def join(self, evt):
        if len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp join <invite link>`")
        elif not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        regex = re.compile(r"(?:https?://)?t(?:elegram)?\.(?:dog|me)(?:joinchat/)?/(.+)")
        arg = regex.match(evt.args[0])
        if not arg:
            return await evt.reply("That doesn't look like a Telegram invite link.")
        arg = arg.group(1)
        if arg.startswith("joinchat/"):
            invite_hash = arg[len("joinchat/"):]
            try:
                await evt.sender.client(CheckChatInviteRequest(invite_hash))
            except InviteHashInvalidError:
                return await evt.reply("Invalid invite link.")
            except InviteHashExpiredError:
                return await evt.reply("Invite link expired.")
            try:
                updates = evt.sender.client(ImportChatInviteRequest(invite_hash))
            except UserAlreadyParticipantError:
                return await evt.reply("You are already in that chat.")
        else:
            channel = await evt.sender.client.get_entity(arg)
            if not channel:
                return await evt.reply("Channel/supergroup not found.")
            updates = await evt.sender.client(JoinChannelRequest(channel))
        for chat in updates.chats:
            portal = po.Portal.get_by_entity(chat)
            if portal.mxid:
                await portal.create_matrix_room(evt.sender, chat, [evt.sender.mxid])
                return await evt.reply(f"Created room for {portal.title}")
            else:
                await portal.invite_matrix([evt.sender.mxid])
                return await evt.reply(f"Invited you to portal of {portal.title}")

    @command_handler
    async def create(self, evt):
        type = evt.args[0] if len(evt.args) > 0 else "group"
        if type not in {"chat", "group", "supergroup", "channel"}:
            return await evt.reply(
                "**Usage:** `$cmdprefix+sp create ['group'/'supergroup'/'channel']`")
        elif not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        if po.Portal.get_by_mxid(evt.room_id):
            return await evt.reply("This is already a portal room.")

        state = await self.az.intent.get_room_state(evt.room_id)
        title = None
        about = None
        levels = None
        for event in state:
            if event["type"] == "m.room.name":
                title = event["content"]["name"]
            elif event["type"] == "m.room.topic":
                about = event["content"]["topic"]
            elif event["type"] == "m.room.power_levels":
                levels = event["content"]
        if not title:
            return await evt.reply("Please set a title before creating a Telegram chat.")
        elif (not levels or not levels["users"] or self.az.intent.mxid not in levels["users"] or
              levels["users"][self.az.intent.mxid] < 100):
            return await evt.reply(f"Please give "
                                   + f"[the bridge bot](https://matrix.to/#/{self.az.intent.mxid})"
                                   + f" a power level of 100 before creating a Telegram chat.")
        else:
            for user, level in levels["users"].items():
                if level >= 100 and user != self.az.intent.mxid:
                    return await evt.reply(
                        f"Please make sure only the bridge bot has power level above"
                        + f"99 before creating a Telegram chat.\n\n"
                        + f"Use power level 95 instead of 100 for admins.")

        supergroup = type == "supergroup"
        type = {
            "supergroup": "channel",
            "channel": "channel",
            "chat": "chat",
            "group": "chat",
        }[type]

        portal = po.Portal(tgid=None, mxid=evt.room_id, title=title, about=about, peer_type=type)
        try:
            await portal.create_telegram_chat(evt.sender, supergroup=supergroup)
        except ValueError as e:
            return await evt.reply(e.args[0])
        return await evt.reply(f"Telegram chat created. ID: {portal.tgid}")

    @command_handler
    async def upgrade(self, evt):
        if not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        portal = po.Portal.get_by_mxid(evt.room_id)
        if not portal:
            return await evt.reply("This is not a portal room.")
        elif portal.peer_type == "channel":
            return await evt.reply("This is already a supergroup or a channel.")
        elif portal.peer_type == "user":
            return await evt.reply("You can't upgrade private chats.")

        try:
            await portal.upgrade_telegram_chat(evt.sender)
            return await evt.reply(f"Group upgraded to supergroup. New ID: {portal.tgid}")
        except ChatAdminRequiredError:
            return await evt.reply("You don't have the permission to upgrade this group.")
        except ValueError as e:
            return await evt.reply(e.args[0])

    @command_handler
    async def groupname(self, evt):
        if len(evt.args) == 0:
            return await evt.reply("**Usage:** `$cmdprefix+sp groupname <name/->`")
        if not evt.sender.logged_in:
            return await evt.reply("This command requires you to be logged in.")

        portal = po.Portal.get_by_mxid(evt.room_id)
        if not portal:
            return await evt.reply("This is not a portal room.")
        elif portal.peer_type != "channel":
            return await evt.reply("Only channels and supergroups have usernames.")

        try:
            await portal.set_telegram_username(evt.sender,
                                               evt.args[0] if evt.args[0] != "-" else "")
            if portal.username:
                return await evt.reply(f"Username of channel changed to {portal.username}.")
            else:
                return await evt.reply(f"Channel is now private.")
        except ChatAdminRequiredError:
            return await evt.reply(
                "You don't have the permission to set the username of this channel.")
        except UsernameNotModifiedError:
            if portal.username:
                return await evt.reply("That is already the username of this channel.")
            else:
                return await evt.reply("This channel is already private")
        except UsernameOccupiedError:
            return await evt.reply("That username is already in use.")
        except UsernameInvalidError:
            return await evt.reply("Invalid username")

    # endregion
    # region Command-related commands
    @command_handler
    def cancel(self, evt):
        if evt.sender.command_status:
            action = evt.sender.command_status["action"]
            evt.sender.command_status = None
            return evt.reply(f"{action} cancelled.")
        else:
            return evt.reply("No ongoing command.")

    @command_handler
    def unknown_command(self, evt):
        return evt.reply("Unknown command. Try `$cmdprefix+sp help` for help.")

    @command_handler
    def help(self, evt):
        if evt.is_management:
            management_status = ("This is a management room: prefixing commands "
                                 "with `$cmdprefix` is not required.\n")
        elif evt.is_portal:
            management_status = ("**This is a portal room**: you must always "
                                 "prefix commands with `$cmdprefix`.\n"
                                 "Management commands will not be sent to Telegram.")
        else:
            management_status = ("**This is not a management room**: you must "
                                 "prefix commands with `$cmdprefix`.\n")
        help = """\n
#### Generic bridge commands
**help**   - Show this help message.  
**cancel** - Cancel an ongoing action (such as login).

#### Authentication
**login** <_phone_> - Request an authentication code.  
**logout**          - Log out from Telegram.  
**ping**            - Check if you're logged into Telegram.

#### Initiating chats
**search** [_-r|--remote_] <_query_> - Search your contacts or the Telegram servers for users.  
**pm** <_identifier_>                - Open a private chat with the given Telegram user. The
                                       identifier is either the internal user ID, the username or
                                       the phone number.  
**join** <_link_>                    - Join a chat with an invite link.  
**create** [_type_]                  - Create a Telegram chat of the given type for the current
                                       Matrix room. The type is either `group`, `supergroup` or
                                       `channel` (defaults to `group`).

#### Portal management  
**upgrade**                - Upgrade a normal Telegram group to a supergroup.  
**invitelink**             - Get a Telegram invite link to the current chat.  
**deleteportal**           - Forget the current portal room. Only works for group chats; to delete
                             a private chat portal, simply leave the room.  
**groupname** <_name_|`-`> - Change the username of a supergroup/channel. To disable, use a dash
                             (`-`) as the name.
"""
        return evt.reply(management_status + help)

    # endregion
    # endregion
